safir/src/safir/schema_manager/_pydantic_schema_manager.py

```diff
@@ -0,0 +1,229 @@
+        schemas registered during testing and staging don't affect the
+        compatibility continuity of a production subject.
+
+        For production, it's best to not set a suffix.
+    """
+
+    def __init__(
+        self,
+        *,
+        registry: AsyncSchemaRegistryClient,
+        suffix: str = "",
+    ) -> None:
+        self._registry = registry
+        self._serializer = AsyncAvroMessageSerializer(self._registry)
+        self._suffix = suffix
+
+        self._logger = logging.getLogger(__name__)
+
+        # A mapping of subjects to registered schema ids.
+        self._subjects_to_ids: dict[str, int] = {}
+
+    async def register_model(
+        self,
+        model: type[AvroBaseModel],
+        compatibility: SchemaRegistryCompatibility | None = None,
+    ) -> SchemaInfo:
+        """Register the model with the registry.
+
+        * The model's schema will be checked for compatibility with the latest
+          version in the registry
+        * An exception will be raised if it's not
+
+        Parameters
+        ----------
+        model
+            The model to register.
+        compatibility
+            The registry `compatibility type <https://docs.confluent.io/platform/current/schema-registry/fundamentals/schema-evolution.html#compatibility-types>`__
+            for this model. If not provided, it will default to the default
+            type configured in the registry.
+        """
+        subject = self._get_subject(model)
+        if compatibility is not None:
+            await self._registry.update_compatibility(
+                subject=subject,
+                level=compatibility,
+            )
+
+        # If a schema has been successfully registered before, and thus has a
+        # version in the registry, trying to create a new version of the schema
+        # in the registry will always succeed and return the schema id of the
+        # previously registered version, regardless of it's compatibility
+        # with the latest version. This can happen when we're trying to "roll
+        # back" a schema.
+        #
+        # In these cases, we want to raise an error if the candidate schema is
+        # incompatible. The only way to accomplish this is to explicitly check
+        # compatibility with the latest version in the registry.
+        #
+        # NOTE: This also doesn't update the 'latest' version to the candidate
+        # schema after a register call of the candidate schema. "latest"
+        # remains at the last registered non-repeat version.
+        schema = model.avro_schema_to_python()
+        result = await self._registry.test_compatibility(
+            subject=subject, schema=schema, verbose=True
+        )
+        if not isinstance(result, dict):
+            raise TypeError(
+                "There seems to be a bug in the python-schema-registry-client"
+                " library!"
+            )
```

Comment:
This is here just to satisfy the type checker. The [test_compatibility](https://github.com/marcosschroh/python-schema-registry-client/blob/09ab31bbdf7bda776e5d344ccf277af9b1350881/schema_registry/client/client.py#L670) method returns a `bool` if `verbose == False`, and a `dict` if `verbose == True`. They could've used `@overload`s or something to make it typecheck nicely, but they didn't :)

Is there a nicer way to satisfy the typechecker here?  Should I just have the error message say "This is only here to satisfy the type checker?"

Reply:
Because it's not properly typed upstream, I am still a bit worried that we might miss an upstream semantic change and the exception might actually fire, so I'm not sure it's only to satisfy the type checker. (It's checking that the contract hasn't changed because the type checker can't do that for us because upstream doesn't have proper typing, or at least that's how I think about it.)

Maybe something like:
```python
if not isinstance(result, dict) or "is_compatible" not in result:
    raise RuntimeError(f"Unexpected result from test_compatibility: {result!s}")
```